web/src/questions-utils.js

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) [2022] SUSE LLC
+ *
+ * All Rights Reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of version 2 of the GNU General Public License as published
+ * by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, contact SUSE LLC.
+ *
+ * To contact SUSE LLC about this file by physical or electronic mail, you may
+ * find current contact information at www.suse.com.
+ */
+
+import React from "react";
+import Popup from "./Popup";
+import { partition } from "./utils";
+
+/**
+ * Returns given text capitalized
+ *
+ * @param {String} text - string to be capitalized
+ * @returns {String} capitalized text
+ */
+const label = text => `${text[0].toUpperCase()}${text.slice(1)}`;
+
+/**
+ * Returns a the options of given question as collection of Popup.Actions, using the defaultOption
+ * as the Popup.PrimaryAction
+ *
+ * TODO: Make it a component?
```

Comment:
I know, it should be not a `TODO` but a component directly. What is more, I did the component too.

<details>
<summary>Click to show/hide the <em>QuestionActoins.jsx</em> component code</summary>

---

```jsx
/*
 * Copyright (c) [2022] SUSE LLC
 *
 * All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of version 2 of the GNU General Public License as published
 * by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, contact SUSE LLC.
 *
 * To contact SUSE LLC about this file by physical or electronic mail, you may
 * find current contact information at www.suse.com.
 */

import React from "react";
import Popup from "./Popup";
import { partition } from "./utils";

const label = action => `${action[0].toUpperCase()}${action.slice(1)}`;

export default function QuestionActions({ actions, defaultAction, actionCallback, conditions }) {
  let [[primaryAction], secondaryActions] = partition(actions, a => a === defaultAction);

  // Ensure there is always a primary action
  if (!primaryAction) [primaryAction, ...secondaryActions] = secondaryActions;

  return (
    <>
      <Popup.PrimaryAction
        key={primaryAction}
        onClick={() => actionCallback(primaryAction)}
        isDisabled={conditions?.disable?.[primaryAction]}
      >
        {label(primaryAction)}
      </Popup.PrimaryAction>
      {
        secondaryActions.map(action =>
          <Popup.SecondaryAction
            key={action}
            onClick={() => actionCallback(action)}
            isDisabled={conditions?.disable?.[action]}
          >
            {label(action)}
          </Popup.SecondaryAction>
        )
      }
    </>
  );
}
```
</details>

But due to the way how the `Popup` component works, it should not include the `<Popup.Actions>` because it is the component that [`Popup` uses to distinguish/extract the actions from the content](https://github.com/yast/d-installer/blob/66faebebe4f413eff704cc3171debc0c5d945e87/web/src/Popup.jsx#L197)

Thus, when using it as it is

```jsx

render (
  <Popup>
    <Text>The question text</Text>
    <QuestionActions />
  </Popup>
)
```

it does not work as expected because `<Popup.Actions>` is not an immediate child of Popup anymore :man_shrugging: 

Changing it to 

```jsx
render (
  <Popup>
    <Text>The question text</Text>
    <Popup.Actions>
      <QuestionActions />
    </Popup.Actions>
  </Popup>
)
```

will work and even we can argue that it is the expected. However, I decided to push it as it is now (before writing that explanation) because

* it renders the `Popup.Actions` as an immediate child of `Popup`

  ```jsx
  return (
    <Popup>
      <Text>The Question</Text>
      { buildQuestionActions(question, actionCallback) }
    </Popup>
  );
```
* as usual, I'd like to hear opinions from you

That said, at this moment I'd vote for making it a component. I don't know what I'll think tomorrow :stuck_out_tongue_winking_eye:

Reply:
Having that `<QuestionActions>` component makes a lot of sense to me. Now we are using it only for the generation of actions in the GenericQuestion. But, actually, we can use it for the LuksActivationQuestion, too.

Reply:
We're using it (actually its _method version_) in the LuksActivationQuestion already :) 

My doubt with the component vs the method was simply because is a bit weird to have the `QuestionAction` components wrapped inside de `Popup.Actions` (a requirement by the `Popup` design. 

A _dumb_ doubt honestly, because there is nothing wrong having either, whatever element inside the `Popup.Actions` or the _Popup_ shortcuts actions (just `Button`s) outside the `Popup.Actions`_wrapper_

In short, I'll move `buildQuestionActions` to `QuestionActions`. Everything will have more sense once we start "grouping" components instead of having all of them in _src/_